pkg/loop/internal/pb/chain_reader.proto

```diff
@@ -9,23 +9,51 @@ import "google/protobuf/empty.proto";
 
 service ChainReader {
   rpc GetLatestValue (GetLatestValueRequest) returns (GetLatestValueReply) {}
+  rpc QueryKey(QueryKeyRequest) returns (QueryKeyReply) {}
   rpc Bind(BindRequest) returns (google.protobuf.Empty) {}
 }
 
 // GetLatestValueRequest has arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.GetLatestValue].
 message GetLatestValueRequest {
-  string contractName = 1;
+  string contract_name = 1;
   string method = 2;
   VersionedBytes params = 3;
 }
 
+// QueryKeyRequest has arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.QueryKey].
+message QueryKeyRequest {
+  string contract_name = 1;
+  QueryKeyFilter filter = 2;
+  LimitAndSort limit_and_sort = 3;
+}
+
+// BindRequest has arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.Bind].
+message BindRequest {
+  repeated BoundContract bindings = 1;
+}
+
 // GetLatestValueReply has return arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.GetLatestValue].
 message GetLatestValueReply {
   VersionedBytes retVal = 1;
 }
 
-message BindRequest {
-  repeated BoundContract bindings = 1;
+// QueryKeyReply has return arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.QueryKey].
+message QueryKeyReply {
+  repeated Sequence sequences = 1;
+}
+
+// Head is gRPC adapter for the Head struct [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.Head].
+message Head {
+  string identifier = 1;
+  bytes hash = 2;
```

Comment:
are there any assumptions in the backend about the length of this? if so, add a comment here about it.

for example in ccip, all the hashes are 32 bytes but since grpc doesn't constrain the length, we just document the expectation next to the definition.

Reply:
All of details like these should get validated in the chain specific implementation since that is always the destination or origin. I don't think that there is a universal fixed length that can work here